File: python/test/test_csv.py
#!/usr/bin/env python3
"""
Test mpc_designation.py against CSV file of known conversions.

Usage: python test_csv.py <csv_file> [max_errors]
"""

import sys
import os
import time

# Add src directory to path for importing mpc_designation
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from mpc_designation import convert_simple, MPCDesignationError


def run_tests(csv_file: str, max_errors: int = 100) -> bool:
    """Run tests from CSV file and report results."""
    total = 0
    passed = 0
    failed = 0
    errors = []

    start_time = time.time()

    with open(csv_file, 'r') as fp:
        # Skip header line
        next(fp)

        for line in fp:
            total += 1
            line = line.strip()
            if not line:
                continue

            parts = line.split(',')
            unpacked = parts[0]
            expected_packed = parts[1]

            try:
                got_packed = convert_simple(unpacked)
            except MPCDesignationError as e:
                failed += 1
                if len(errors) < max_errors:
                    errors.append((unpacked, f"ERROR: {e}", expected_packed))
                continue

            if got_packed != expected_packed:
                failed += 1
                if len(errors) < max_errors:
                    errors.append((unpacked, got_packed, expected_packed))
            else:
                passed += 1

            # Progress indicator every 100,000 entries
            if total % 100000 == 0:
                print(f"Processed {total} entries...")

    elapsed = time.time() - start_time

    print()
    print("=== Test Results ===")
    print(f"Total:  {total}")
    print(f"Passed: {passed}")
    print(f"Failed: {failed}")
    print(f"Time:   {elapsed*1000:.0f}ms ({total/elapsed:.1f} entries/sec)")
    print()

    if failed > 0:
        print(f"=== First {len(errors)} failures ===")
        print(f"{'Input':<25} {'Got':<15} {'Expected':<15}")
        print("-" * 60)
        for inp, got, expected in errors:
            print(f"{inp:<25} {got:<15} {expected:<15}")

    return failed == 0


def main():
    if len(sys.argv) < 2:
        print("Usage: test_csv.py <csv_file> [max_errors]", file=sys.stderr)
        sys.exit(1)

    csv_file = sys.argv[1]
    max_errors = int(sys.argv[2]) if len(sys.argv) > 2 else 100

    success = run_tests(csv_file, max_errors)
    sys.exit(0 if success else 1)


if __name__ == '__main__':
    main()

File: js/test/test_csv.js
#!/usr/bin/env node
/**
 * Test mpc_designation.js against CSV file of known conversions.
 *
 * Usage: node test_csv.js <csv_file> [max_errors]
 */

'use strict';

const fs = require('fs');
const readline = require('readline');
const path = require('path');

const { convertSimple, MPCDesignationError } = require('../src/mpc_designation');

async function runTests(csvFile, maxErrors = 100) {
    let total = 0;
    let passed = 0;
    let failed = 0;
    const errors = [];

    const startTime = Date.now();

    const fileStream = fs.createReadStream(csvFile);
    const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
    });

    let isFirstLine = true;

    for await (const line of rl) {
        if (isFirstLine) {
            isFirstLine = false;
            continue; // Skip header
        }

        const trimmedLine = line.trim();
        if (!trimmedLine) {
            continue;
        }

        total++;
        const parts = trimmedLine.split(',');
        const unpacked = parts[0];
        const expectedPacked = parts[1];

        let gotPacked;
        try {
            gotPacked = convertSimple(unpacked);
        } catch (e) {
            failed++;
            if (errors.length < maxErrors) {
                errors.push([unpacked, `ERROR: ${e.message}`, expectedPacked]);
            }
            continue;
        }

        if (gotPacked !== expectedPacked) {
            failed++;
            if (errors.length < maxErrors) {
                errors.push([unpacked, gotPacked, expectedPacked]);
            }
        } else {
            passed++;
        }

        // Progress indicator every 100,000 entries
        if (total % 100000 === 0) {
            process.stdout.write(`Processed ${total} entries...\n`);
        }
    }

    const elapsed = Date.now() - startTime;

    console.log();
    console.log('=== Test Results ===');
    console.log(`Total:  ${total}`);
    console.log(`Passed: ${passed}`);
    console.log(`Failed: ${failed}`);
    console.log(`Time:   ${elapsed}ms (${(total / (elapsed / 1000)).toFixed(1)} entries/sec)`);
    console.log();

    if (failed > 0) {
        console.log(`=== First ${errors.length} failures ===`);
        console.log(`${'Input'.padEnd(25)} ${'Got'.padEnd(15)} ${'Expected'.padEnd(15)}`);
        console.log('-'.repeat(60));
        for (const [inp, got, expected] of errors) {
            console.log(`${inp.padEnd(25)} ${got.padEnd(15)} ${expected.padEnd(15)}`);
        }
    }

    return failed === 0;
}

async function main() {
    const args = process.argv.slice(2);

    if (args.length < 1) {
        console.error('Usage: node test_csv.js <csv_file> [max_errors]');
        process.exit(1);
    }

    const csvFile = args[0];
    const maxErrors = args.length > 1 ? parseInt(args[1], 10) : 100;

    const success = await runTests(csvFile, maxErrors);
    process.exit(success ? 0 : 1);
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});
